### DISCUSSION TITLE
Bayesian change detection models? #feature

### DISCUSSION THREAD

**Discussion by hyiltiz** [iSoron/uhabits#311] [Feature Requests]
I checked out the Wikipedia page on exponential smoothing, and was quite disappointed by the algorithm, since it does not take into account many possible factors, such as habit difficulty, possible causes of failure etc. Consider I would like to try to sleep before midnight, then sleeping at 1 is much better than 4, but the current model does not address that. Also, consider I'd like to firm a new habit of drinking tea in the morning, but sometimes I couldn't help and end up drinking coffee, or even coke! These different alternatives could represent very different utilities for me, so would make sense to treat coke and coffee differently.

Now, with change detection models, we can assume possibilities (difficulty measure) of each alternative happening, and their preferred value (to our personal well-beings). We can then predict the possibility (*score* in the software) of the event that we'd like to form into habit happening next times given all history, their difficulty levels (it could change and the model can estimate that, which is part of the model) etc. 

This is about the leading edge of science in this field. I could help in algorithm (already have a package in Marian but I know my languages) if you guys think this is a cool idea, and would like to try out. We could put this option as advanced, and give a bit introduction about model input parameters before the user turns it on.

See paper Adams and McKay (2007) for more details.

**Comment by iSoron** [comment]
@hyiltiz Sounds interesting, but what additional input parameters would this model need? All the data we currently have is whether the use performed the habit or not on a given day.

**Comment by hyiltiz** [comment]
The input parameters would depend on the type of habit since that would characterize the model. For each habit, statistically speaking, we could see it as the preferred events of all the possible events that could happen, and each events have their outcomes (utilities). Perhaps this would be easier if I follow up with my previous two examples:

_Habit 1_: Sleep before midnight (the sooner the better)
I could possibly go to sleep at many different hours (possible events), e.g. 00:00, 01:00, 02:15 etc., and for simplicity, let's assume that the possibilities of each of them happening are equal (a uniform probability distribution). Now, for my own well-being, I would prefer to sleep as early as possible, so I would give more points (utility) if I sleep sooner (than later). 
For this habit, the inputs could be: the range of possible hours of going to sleep (00:00-04:00), their possibilities (uniform), and my utility (e.g. exponential decay). With these inputs, we could set up an updating rule that only calculates several numbers everytime, and the user selects one of the possible events from the range (00:00-04:00) rather than simply putting a checkmark.

_Habit 2_: Drinking tea in the morning
This is a simpler case since it is discrete. The user would have to specify the following table as inputs.

| Possible events | tea | milk | coffee | beer | water |
| ---- | --- | --- | --- | ---- | -------- |
| Possibilities | 0.05 | 0.2 | 0.6 | 0.02 |  0.1 |
| Utilities | .6 | 0.2 | 0.05 | 0.05 |  0.1 |

Then, we could even monitor when they are *not* likely to form a habit, and their habit is *endangered*, so to speak (change detection), in addition to just simply prediction if it is happening (score). 

I understand that this is quite advanced as most users wouldn't be able to understand the model or how it works, unless explained very carefully in clear plain language. 

Here is the link for the paper (was on phone previously): https://arxiv.org/abs/0710.3742

**Comment by deutrino** [comment]
What about a model like 

* did habit
* mostly did habit 
* did habit a little 
* didn't do habit

For me, abstaining from news and social media three days a week is a habit I'm working on, but if I cheat even a little, for example by checking one social media account, then it becomes "well I can't check the habit for today, might as well party right? ::loads politico.com::" and that's not necessarily the best outcome 😂

**Comment by extensibleseth** [comment]
Could you define a habit quality/value range slider?

**Comment by hyiltiz** [comment]
I am willing to write an interface that handles the internal logic (the statistical modeling) correctly for this, if @iSoron or someone else can show me an example. I have no experience and willingness to write Java or Kotlin at all, but I "speak" a dozen of them so shouldn't be hard especially it will mostly be math. The function will be something like this:

```
function updateHabitProgress(
  habitID, 
  habitAccomplishmentProgressCurrent, # a value ranging from 0 to 1
  [habitProfile], # an array user-specified events that form the habit, with
    difficulties and importance for each event evaluated by the user (e.g. using
    a slider); an example would be [("beer", 0.9, 0.1), ("wine", 0.3,0.8)] for a
    "Quit drinking" habit, specifying that to quit drinking, the user need to
    quit beer (very hard for the user),but not as imoprtant as quitting wine
    (not quite hard, specified by 0.3) yet very important for the habit
    (specified by 0.8)
  userResponse, # a value ranging from 0 to 1 (e.g. input from a slider)
  [habitAccomplishmentProgressHistory] # this is an *optional* array input
  )

  ...

  return habitAccomplishmentProgressNew 
```

Arguably, `habitID` can provide a handle that lookup `habitProfile` and `habitAccomplishmentProgressHistory`, but I'd rather not deal with internals such as database lookup etc. So, `habitID` is really there for compatibility reasons so it might be useful in future. Others can use this function as a library, and wrap it with something more native to the project.

Anyone interested?

**Comment by iSoron** [comment]
@deutrino @extensibleseth We are currently working on numerical habits (e.g., how many miles did you run today?) and after that works well enough, the plan is to add support to other variations (e.g. How well did you do something? Good, acceptable, poor?). Behind the scenes, this just assign a numerical value for each option, say, from 1 to 10.

What @hyiltiz is suggesting seems to be different, though. We could certainly make the app accept a customized set of answers (e.g. What did you drink today?  Milk, Tea, Coffee?), let the user assign a certain score (or utility) to each answer (e.g. Milk: 100 points, Tea: 50 points, Coffee: 10 points), then compute the current score based on these numbers. What is not clear to me is why do we need the users to also specify the difficulty. If the users assign very high utility to a certain label, but never chooses it, it should be immediately clear that the label is difficult.

**Comment by hyiltiz** [comment]
I think difficulty could be subjective and objective as well, and when it comes to shaping habits, those who have "strong will power" might even tame "difficult" habits (subjective or objective). I guess, as for subjective difficulty, we can ignore it or only use it as prior (a starting point), and infer/compute objective difficulty based on the habit history, as @iSoron suggested.

**Comment by hyiltiz** [comment]
Just wanted to follow up. What do you guys think? Shall we drop this for now? Otherwise I could write up the function I mentioned above and send a pull request.
